src/cmd/tools/verify_data_files/main/main.go

```diff
@@ -0,0 +1,473 @@
+			}
+
+			shardFileSets, err := fs.DataFiles(filePathPrefix,
+				ident.StringID(namespace), uint32(shardID))
+			if err != nil {
+				log.Fatal("could not list shard dir file setes",
+					zap.String("shardPath", shardPath), zap.Error(err))
+			}
+
+			log.Debug("discovered shard file sets",
+				zap.String("namespace", namespace),
+				zap.String("namespacePath", namespacePath),
+				zap.Int("shardID", shardID),
+				zap.Any("fileSets", shardFileSets))
+			fileSetFiles = append(fileSetFiles, shardFileSets...)
+		}
+	}
+
+	// Sort by time in reverse (usually want to fix latest files first and
+	// can stop once done with fail-fast).
+	log.Info("sorting file sets", zap.Int("numFileSets", len(fileSetFiles)))
+	sort.Slice(fileSetFiles, func(i, j int) bool {
+		return fileSetFiles[i].ID.BlockStart.After(fileSetFiles[j].ID.BlockStart)
+	})
+
+	log.Info("verifying file sets", zap.Int("numFileSets", len(fileSetFiles)))
+	for _, fileSet := range fileSetFiles {
+		log.Info("verifying file set file", zap.Any("fileSet", fileSet))
+		if err := verifyFileSet(verifyFileSetOptions{
+			filePathPrefix: filePathPrefix,
+			bytesPool:      bytesPool,
+			fileSet:        fileSet,
+			fixDir:         opts.fixDir,
+			fixInvalidIDs:  opts.fixInvalidIDs,
+			fixInvalidTags: opts.fixInvalidTags,
+		}, log); err != nil {
+			log.Error("file set file failed verification",
+				zap.Error(err),
+				zap.Any("fileSet", fileSet))
+
+			if opts.failFast {
+				log.Fatal("aborting due to fail fast set")
+			}
+		}
+	}
+}
+
+func dirFiles(dirPath string) ([]string, error) {
+	dir, err := os.Open(dirPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not open dir: %v", err)
+	}
+
+	defer dir.Close()
+
+	stat, err := dir.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("could not stat dir: %v", err)
+	}
+	if !stat.IsDir() {
+		return nil, fmt.Errorf("path is not a directory: %s", dirPath)
+	}
+
+	entries, err := dir.Readdirnames(-1)
+	if err != nil {
+		return nil, fmt.Errorf("could not read dir names: %v", err)
+	}
+
+	results := make([]string, 0, len(entries))
+	for _, p := range entries {
+		if p == "." || p == ".." || p == "./.." || p == "./" || p == "../" || p == "./../" {
+			continue
+		}
+		results = append(results, p)
+	}
+	return results, nil
+}
+
+type verifyFileSetOptions struct {
+	filePathPrefix string
+	bytesPool      pool.CheckedBytesPool
+	fileSet        fs.FileSetFile
+
+	fixDir         string
+	fixInvalidIDs  bool
+	fixInvalidTags bool
+}
+
+func verifyFileSet(
+	opts verifyFileSetOptions,
+	log *zap.Logger,
+) error {
+	fsOpts := fs.NewOptions().SetFilePathPrefix(opts.filePathPrefix)
+	reader, err := fs.NewReader(opts.bytesPool, fsOpts)
+	if err != nil {
+		return err
+	}
+
+	fileSet := opts.fileSet
+
+	openOpts := fs.DataReaderOpenOptions{
+		Identifier:  fileSet.ID,
+		FileSetType: persist.FileSetFlushType,
+	}
+
+	err = reader.Open(openOpts)
+	if err != nil {
+		return err
+	}
+
+	defer reader.Close()
```

Comment:
Should this be deferred earlier before the error above?

Reply:
Close() is only valid after an Open(), hence doing it below the return